impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java

```diff
@@ -220,7 +220,10 @@ public OnHeapStore(final Configuration<K, V> config, TimeSource timeSource, Copi
           }
 
           if (updateAccess) {
-            setAccessTimeAndExpiry(key, mappedValue, now);
+            if (setAccessTimeAndExpiryThenReturnMapping(key, mappedValue, now) == null) {
```

Comment:
Still believe this whole block can be hidden inside a method. Also your first version had many more locations where equivalent code was called, but you changed to a version that no longer notifies.
I believe the initial version was more _correct_  just needed some refactoring to extract duplication.